File: functions/src/contact/models/contact_model.py
import json
from typing import Any, Dict, List, Optional
from src.core.models.phone_model import Phone


class ContactModel:
    def __init__(
        self,
        name: str,
        phone_numbers: List[Phone],
        id: Optional[str] = None,
        photo: Optional[str] = None,  # New photo field
    ):
        self.name = name
        self.phone_numbers = phone_numbers
        self.id = id
        self.photo = photo  # Assign the photo field

    def copy_with(
        self,
        name: Optional[str] = None,
        phone_numbers: Optional[List[Phone]] = None,
        id: Optional[str] = None,
        photo: Optional[str] = None,  # New photo field for copy
    ) -> "ContactModel":
        return ContactModel(
            name=name or self.name,
            phone_numbers=phone_numbers or self.phone_numbers,
            id=id if id is not None else self.id,
            photo=photo if photo is not None else self.photo,  # Copy photo
        )

    def to_map(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "phoneNumbers": [phone.to_map() for phone in self.phone_numbers],
            "id": self.id,
            "photo": self.photo,  # Include photo in map
        }

    def to_json(self) -> str:
        return json.dumps(self.to_map())

    @staticmethod
    def from_json(data: str) -> "ContactModel":
        return ContactModel.from_map(json.loads(data))

    @staticmethod
    def from_map(data: Dict[str, Any]) -> "ContactModel":
        return ContactModel(
            name=data.get("name", ""),
            phone_numbers=[
                Phone.from_map(phone) for phone in data.get("phoneNumbers", [])
            ],
            id=data.get("id"),
            photo=data.get("photo"),  # Extract photo from map
        )

File: lib/src/contact/models/contact_model.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter_contacts/contact.dart' as flutter_contacts;
import '../../../core/models/phone_model.dart';

class ContactModel {
  final String name;
  final List<PhoneModel> phoneNumbers;
  final String? id;
  final String? photo;

  ContactModel({
    required this.name,
    required this.phoneNumbers,
    this.id,
    this.photo,
  });

  factory ContactModel.fromContact(flutter_contacts.Contact contact) {
    return ContactModel(
      name: contact.name.first,
      phoneNumbers: contact.phones
          .map(
            (phone) => PhoneModel(
              isoCode: "",
              dialCode: "",
              phoneNumber: phone.number,
            ),
          )
          .toList(),
    );
  }

  ContactModel copyWith({
    String? name,
    List<PhoneModel>? phoneNumbers,
    String? id,
    String? photo,
  }) {
    return ContactModel(
      name: name ?? this.name,
      phoneNumbers: phoneNumbers ?? this.phoneNumbers,
      id: id ?? this.id,
      photo: photo ?? this.photo,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'phoneNumbers': phoneNumbers.map((x) => x.toMap()).toList(),
      'id': id,
      'photo': photo,
    };
  }

  factory ContactModel.fromMap(Map<String, dynamic> map) {
    return ContactModel(
      name: map['name'] ?? '',
      phoneNumbers: List<PhoneModel>.from(
          map['phoneNumbers']?.map((x) => PhoneModel.fromMap(x))),
      id: map['id'],
      photo: map['photo'],
    );
  }

  String toJson() => json.encode(toMap());

  factory ContactModel.fromJson(String source) =>
      ContactModel.fromMap(json.decode(source));

  @override
  String toString() {
    return 'ContactModel(name: $name, phoneNumbers: $phoneNumbers, id: $id, photo: $photo)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is ContactModel &&
        other.name == name &&
        listEquals(other.phoneNumbers, phoneNumbers) &&
        other.id == id &&
        other.photo == photo;
  }

  @override
  int get hashCode {
    return name.hashCode ^ phoneNumbers.hashCode ^ id.hashCode ^ photo.hashCode;
  }

  static ContactModel empty() => ContactModel(
        name: "",
        phoneNumbers: [],
      );
}
